staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go

```diff
@@ -275,6 +276,25 @@ func finishRequest(ctx context.Context, fn resultFunc) (result runtime.Object, e
 	case p := <-panicCh:
 		panic(p)
 	case <-ctx.Done():
+		defer func() {
+			go func() {
+				waitStarted := time.Now()
+
+				panicd := false
+				var requestErr error
+				select {
```

Comment:
So if the call truly hangs forever, with `time.sleep(365*24*time.hour)`, this would stack up a second set of infinite go funcs for logging purposes?

It might be easier to judge correctness here by 
1. creating  struct containing all the various channels so you can comment and give a better than `ch`
2. avoid turning buffered channels into unbuffered channels (it's not obvious why a consumer must exist and should block our producer thread)
2. factoring with a method on that struct to avoid giant inline
3. adding a timeout on waiting for the timeout to finish. Even something order of 30 minutes looks lower risk

Reply:
So I took a stab at refactoring the `finishRequest` function:
- removed the function to its own package
- use a single channel to propagate the result from the child goroutine to the parent.

This refactor allows me to inject fakes so I can test the post-timeout behavior. I wrote unit tests to verify that the post-timeout `receiver` goroutine we spin up does not block, it waits for the child goroutine to return a value. So now the post-timeout logic is properly tested and verified.

@deads2k @sttts  this refactor seems out of scope of this PR, let me know if you want me to open a new PR with just the refactor.


> avoid turning buffered channels into unbuffered channels (it's not obvious why a consumer must exist and should block our producer thread)

Now, we spin up a post-timeout receiver goroutine that is waiting to receive, so there is no need to make the channel buffered. Also, the test I wrote ensures that the child goroutine does not block on sending after the request times out.